pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3407 @@
+        # expression, and if a is polynomial and b is a positive
+        # integer, it is also polynomial.  While we would like to just
+        # call this a non-polynomial expression, these exceptions occur
+        # too frequently (and in particular, a**2)
+        l,r = result
+        if r == 0:
+            if l == 0:
+                return 0
+            # NOTE: use value before int() so that we don't
+            #       run into the disabled __int__ method on
+            #       NumericValue
+            exp = value(self._args_[1], exception=False)
+            if exp is None:
+                return None
+            if exp == int(exp):
+                if l is not None and exp > 0:
+                    return l * exp
+                elif exp == 0:
+                    return 0
+        return None
+
+    def _is_fixed(self, args):
+        assert(len(args) == 2)
+        if not args[1]:
+            return False
+        return args[0] or value(self._args_[1]) == 0
+
+    def _precedence(self):
+        return PowExpression.PRECEDENCE
+
+    def _apply_operation(self, result):
+        _l, _r = result
+        return _l ** _r
+
+    def getname(self, *args, **kwds):
+        return 'pow'
+
+    def _to_string(self, values, verbose, smap, compute_values):
+        if verbose:
+            return "{0}({1}, {2})".format(self.getname(), values[0], values[1])
+        return "{0}**{1}".format(values[0], values[1])
+
+
+class NPV_PowExpression(PowExpression):
+    __slots__ = ()
+
+    def is_potentially_variable(self):
+        return False
+
+
+class ProductExpression(ExpressionBase):
+    """
+    Product expressions::
+
+        x*y
+    """
+
+    __slots__ = ()
+    PRECEDENCE = 4
+
+    def _precedence(self):
+        return ProductExpression.PRECEDENCE
+
+    def _compute_polynomial_degree(self, result):
+        # NB: We can't use sum() here because None (non-polynomial)
+        # overrides a numeric value (and sum() just ignores it - or
+        # errors in py3k)
+        a, b = result
+        if a is None or b is None:
+            return None
+        else:
+            return a + b
+
+    def getname(self, *args, **kwds):
+        return 'prod'
+
+    def _apply_operation(self, result):
+        _l, _r = result
+        return _l * _r
+
+    def _to_string(self, values, verbose, smap, compute_values):
+        if verbose:
+            return "{0}({1}, {2})".format(self.getname(), values[0], values[1])
+        if values[0] == "1" or values[0] == "1.0":
+            return values[1]
+        if values[0] == "-1" or values[0] == "-1.0":
+            return "- {0}".format(values[1])
+        return "{0}*{1}".format(values[0],values[1])
+
+
+class NPV_ProductExpression(ProductExpression):
+    __slots__ = ()
+
+    def is_potentially_variable(self):
+        return False
+
+
+class MonomialTermExpression(ProductExpression):
+    __slots__ = ()
+
+
+class ReciprocalExpression(ExpressionBase):
```

Comment:
I think this should be returned to DivisionExpression.  I am not convinced that adding to the number of nodes in a tree is a good thing.